File: src/main/programs/array/TopKFrequent.java
package array;

import java.util.*;

public class TopKFrequent {
    public static int[] topKFrequent(int[] nums, int k) {
        // 1. Count frequencies
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }

        // 2. Min-heap to keep top k elements
        PriorityQueue<Integer> heap = new PriorityQueue<>(
                (a, b) -> freqMap.get(a) - freqMap.get(b)
        );

        for (int num : freqMap.keySet()) {
            heap.offer(num);
            if (heap.size() > k) {
                heap.poll(); // Remove least frequent
            }
        }

        // 3. Convert heap to array
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = heap.poll();
        }
        return result;
    }

    public List<Integer> topKFrequents(int[] nums, int k) {
        // 1. Frequency map
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);

        // 2. Bucket array: index = frequency
        List<Integer>[] buckets = new List[nums.length + 1];
        for (int num : freqMap.keySet()) {
            int freq = freqMap.get(num);
            if (buckets[freq] == null) buckets[freq] = new ArrayList<>();
            buckets[freq].add(num);
        }

        // 3. Extract top k from highest frequency buckets
        List<Integer> result = new ArrayList<>();
        for (int i = buckets.length - 1; i >= 0 && result.size() < k; i--) {
            if (buckets[i] != null) result.addAll(buckets[i]);
        }
        return result.subList(0, k);
    }

    public static void main(String[] args) {
        int[] nums = {1, 1, 1, 2, 2, 3};
        int k = 2;
        System.out.println(Arrays.toString(topKFrequent(nums, k))); // [2, 1] or [1, 2]
    }
}

File: src/main/programs/array/TopKFrequent.kt
package array

import java.util.PriorityQueue

fun topKFrequentPQ(nums:IntArray,k:Int):IntArray{
val frequencyMap= mutableMapOf<Int,Int>()
    nums.forEach { num->
        frequencyMap[num] = frequencyMap.getOrDefault(num,0)+1
    }

    val pq=PriorityQueue<Map.Entry<Int,Int>>{a,b->
        b.value-a.value
    }
    pq.addAll(frequencyMap.entries)
    return IntArray(k){pq.poll().key}
}

fun findTopKFrequent(nums: IntArray, k: Int): IntArray {
    return nums.toList()
        .groupingBy { it }
        .eachCount()
        .entries
        .sortedByDescending { it.value }
        .take(k)
        .map { it.key }
        .toIntArray()
}



fun main(){
    val num= intArrayOf(1,1,1,2,2,3)
    val k=2
    //val result= topKFrequentPQ(num,k)
    val result= findTopKFrequent(num,k)
    println("Top $k frequent element: ${result.contentToString()}")
}